### PR TITLE
add classes for 3 figures, Figure interface and function getInfo

### PR DESCRIPTION
**Pull request by Koliras**

### CODE REVIEW

**Review by yuliiaant — APPROVED** [review]
Good job!

File: src/figuresClasses.ts [hunk 1]
```diff
@@ -1,19 +1,87 @@
-export interface Figure {
-
+function isTrianglePossible(a: number, b: number, c: number): boolean {
+  return a + b > c && a + c > b && b + c > a;
 }
 
-export class Triangle {
+function roundDownFloat(number: number, amountOfDecimals: number = 2): number {
+  return Math.floor(number * 10 ** amountOfDecimals) / 10 ** amountOfDecimals;
```
Comment (Moroz-Dmytro) [inline src/figuresClasses.ts:6]: Can be moved to variable - `10 ** amountOfDecimals`

File: src/figuresClasses.ts [hunk 2]
```diff
@@ -1,19 +1,87 @@
...
 
+export interface Figure {
+  shape: string,
+  color: string,
+  getArea: Function,
 }
 
-export class Circle {
+export class Triangle implements Figure {
+  public shape = 'triangle';
```
Comment (Moroz-Dmytro) [inline src/figuresClasses.ts:16]: Create `enum` for `shape`

File: src/figuresClasses.ts [hunk 3]
```diff
@@ -1,19 +1,87 @@
...
+
+  constructor(
+    public color: string,
```
Comment (Moroz-Dmytro) [inline src/figuresClasses.ts:19]: create `enum` for `color`

File: src/figuresClasses.ts [hunk 4]
```diff
@@ -1,19 +1,87 @@
...
+  }
+
+  getArea(): number {
+    const semiperimeter = (this.a + this.b + this.c) * 0.5;
 
+    const area = (semiperimeter
+      * (semiperimeter - this.a)
+      * (semiperimeter - this.b)
+      * (semiperimeter - this.c)
+    ) ** 0.5;
```
Comment (Moroz-Dmytro) [inline src/figuresClasses.ts:40]: Better to use `Math.sqrt`. Read more [here](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Math/sqrt)

File: src/figuresClasses.ts [hunk 5]
```diff
@@ -1,19 +1,87 @@
...
+
+  getArea(): number {
+    const area = this.height * this.width;
+
+    return roundDownFloat(area);
+  }
+}
 
+export function getInfo(figure: Figure): string {
+  return `A ${figure.color} ${figure.shape} - ${figure.getArea()}`;
```
Comment (Moroz-Dmytro) [inline src/figuresClasses.ts:86]: We can use destructuring and create variable for `${figure.getArea()}` for better readability

File: src/figuresClasses.ts [hunk 6]
```diff
@@ -1,19 +1,87 @@
...
+      throw new Error("Sides of the triangle can't be equal or smaller than 0");
+    }
+
+    if (!isTrianglePossible(a, b, c)) {
+      throw new Error("One side can't be greater than the sum of two others");
+    }
+  }
+
+  getArea(): number {
+    const semiperimeter = (this.a + this.b + this.c) * 0.5;
```
Comment (Moroz-Dmytro) [inline src/figuresClasses.ts:34]:
```suggestion
    const semiperimeter = (this.a + this.b + this.c) / 2;
```

**Review by Moroz-Dmytro — CHANGES_REQUESTED** [review]

File: src/figuresClasses.ts [hunk 7]
```diff
@@ -1,19 +1,99 @@
-export interface Figure {
+function isTrianglePossible(a: number, b: number, c: number): boolean {
+  return a + b > c && a + c > b && b + c > a;
+}
```
Comment (alexander-ignatow) [inline src/figuresClasses.ts:3]: I would place it inside triangle class, as it only deals with triangles

File: src/figuresClasses.ts [hunk 8]
```diff
@@ -1,19 +1,99 @@
...
 
+enum Shape {
+  Triangle = 'triangle',
+  Circle = 'circle',
+  Rectangle = 'rectangle',
 }
 
-export class Triangle {
+type Color = 'red'|'blue'|'green';
```
Comment (alexander-ignatow) [inline src/figuresClasses.ts:17]: What is the reason for using enums and unions together?

File: src/figuresClasses.ts [hunk 9]
```diff
@@ -1,19 +1,99 @@
...
+
+  constructor(
+    public color: Color,
+    public a: number,
+    public b: number,
+    public c: number,
+  ) {
+    if ([a, b, c].some((number) => number <= 0)) {
+      throw new Error("Sides of the triangle can't be equal or smaller than 0");
+    }
```
Comment (alexander-ignatow) [inline src/figuresClasses.ts:36]: 👍

File: src/figuresClasses.ts [hunk 10]
```diff
@@ -1,19 +1,99 @@
...
+    return roundDownFloat(area);
+  }
 }
 
-export function getInfo(figure) {
+export function getInfo(figure: Figure): string {
+  const figureArea = figure.getArea();
+  const { color, shape } = figure;
 
+  return `A ${color} ${shape} - ${figureArea}`;
```
Comment (alexander-ignatow) [inline src/figuresClasses.ts:98]: Consider @Moroz-Dmytro comment on destructing figure

**Review by alexander-ignatow — CHANGES_REQUESTED** [review]
Good! 👍 

Lets fix few more issues

File: src/figuresClasses.ts (hunk 10 above)
Comment (Koliras) [reply src/figuresClasses.ts:98]: @alexander-ignatow But if I destruct a figure, how can I get the area of the figure?

File: src/figuresClasses.ts (hunk 8 above)
Comment (Koliras) [reply src/figuresClasses.ts:17]: If I use enum tests go well but they are showing problems
image

File: src/figuresClasses.ts [hunk 11]
```diff
@@ -1,19 +1,103 @@
...
+    }
+  }
+
+  getArea(): number {
+    const semiperimeter = (this.a + this.b + this.c) / 2;
+
+    const area = Math.sqrt(semiperimeter
+      * (semiperimeter - this.a)
+      * (semiperimeter - this.b)
+      * (semiperimeter - this.c));
```
Comment (Moroz-Dmytro) [inline src/figuresClasses.ts:53]:
```suggestion
    const area = Math.sqrt(semiperimeter
      * (semiperimeter - this.a)
      * (semiperimeter - this.b)
      * (semiperimeter - this.c)
    );
```

**Review by Moroz-Dmytro — APPROVED** [review]
LGTM 👍

**Pull request closed without merging**